### PR TITLE
Improvements to Pyroguard Emberseer (UBRS) script

### PR DESCRIPTION
**Pull request by sveN295**
**Changes proposed:**

- Using doorData to activate doors
- Blackhand Incarcerators are non-hostile until the altar is activated
- Blackhand Incarcerator respawn on wipe
- Blackhand Incarcerator cast Encage Emberseer when not engaged
- Altar only requires one person to channel instead of everyone in the instance (see https://wowwiki.fandom.com/wiki/Pyroguard_Emberseer)

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)
Adresses #19388

**Tests performed:** (Does it build, tested in-game, etc.)
- [x] builds
- [x] tested in-game

### CODE REVIEW

**Comment by Aokromes** [comment]
nice!

**Comment by illfated** [comment]
Looks like **ci/circleci: pch_and_sql** only complains about whitespace.

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp [hunk 1]
```diff
@@ -437,6 +366,7 @@ class npc_blackhand_incarcerator : public CreatureScript
 
         private:
             EventMap _events;
+            InstanceScript* _instance;
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp:369]: where is this member field used ?

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp [hunk 2]
```diff
@@ -341,39 +305,18 @@ class npc_blackhand_incarcerator : public CreatureScript
...
-            if (data == 1 && value == 1)
-            {
-                me->SetImmuneToAll(false);
-                me->InterruptSpell(CURRENT_CHANNELED_SPELL);
-                _events.CancelEvent(EVENT_ENCAGED_EMBERSEER);
-            }
-
-            if (data == 1 && value == 2)
-                _events.ScheduleEvent(EVENT_ENCAGED_EMBERSEER, 1s);
+            DoCast(SPELL_ENCAGE_EMBERSEER);
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp:315]: Never put anything other than assigning class fields in Initialize(). Never cast spells/do strange things in any function called by the constructor.
Spells should be cast in Reset() or JustAppeared()

**Review by jackpoz — CHANGES_REQUESTED** [review]

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp (hunk 1 above)
Comment (sveN295) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp:366]: Good point, forgot to remove it from a previous version where it was used.

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp (hunk 2 above)
Comment (sveN295) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp:312]: Thank you for the feedback, will not do that in the future.

**Review by jackpoz — APPROVED** [review]

**Review by ghost — APPROVED** [review]
Looks quite nice. :-)

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp [hunk 3]
```diff
@@ -253,13 +217,13 @@ class boss_pyroguard_emberseer : public CreatureScript
...
-                            bool _hasAura = true;
+                            // As of Patch 3.0.8 only one person needs to channel the altar
+                            bool _hasAura = false;
                             Map::PlayerList const& players = me->GetMap()->GetPlayers();
                             for (Map::PlayerList::const_iterator itr = players.begin(); itr != players.end(); ++itr)
                                 if (Player* player = itr->GetSource()->ToPlayer())
-                                    if (!player->HasAura(SPELL_EMBERSEER_OBJECT_VISUAL))
-                                        _hasAura = false;
+                                    if (player->HasAura(SPELL_EMBERSEER_OBJECT_VISUAL))
+                                        _hasAura = true;
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp:226]: should there be a break; after this to stop at the first found player ? (with the required { } etc)

**Review by jackpoz — CHANGES_REQUESTED** [review]

**Comment by Rushor** [comment]
@sveN295 anything you need help for or that should be tested or anything you may need improvement for?

**Comment by sveN295** [comment]
@Rushor Thanks for the offer, I would really like if someone can doublecheck the functionality in-game. 

I also have a question:

https://github.com/TrinityCore/TrinityCore/blob/35b001c1d47c737d83679e827fde0badb34680d4/src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp#L152-L156

What is this part of OnGameObjectCreate() for exactly? It seems to be redundant after adding doorData, if i comment it out the door still opens when the boss is killed. Are there any possible side effects of removing this?

I would also really like an explanation for this:

https://github.com/TrinityCore/TrinityCore/blob/35b001c1d47c737d83679e827fde0badb34680d4/src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp#L272-L292

I haven't seen ProcessEvent() anywhere else yet and can't find a documentation about how it works and where the eventId comes from.

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp (hunk 3 above)
Comment (sveN295) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_pyroguard_emberseer.cpp:227]: Yes, you're absolutely right.

**Comment by Jinnaix** [comment]
Holy Cow @jackpoz, your newest codestyle check is ultra picky :D

**Review by jackpoz — APPROVED** [review]

**Comment by jackpoz** [comment]
@Jinnaix : it's just getting enforced, TC had the whitespace rule for many years now ( see https://github.com/TrinityCore/TrinityCore/commit/af2d4f039bd3dc87e8112adf1d8af0744e6ac11d#diff-bdbbdc184d1c03e11833b6cdfd174025 ) 😃

**Comment by illfated** [comment]
[War on white spaces](https://github.com/TrinityCore/TrinityCore/pull/24405/commits/04ab1563c99dcefd1d2515c5d5824931fdbd455e) - that commit name cracked me up. 😄

**Comment by jackpoz** [comment]
@sveN295 is this PR ready to be merged ? did you give a last check after all the changes done during the PR review ?

**Comment by Jinnaix** [comment]
We really lack people that test pull requests and provide feedback :/

**Comment by jackpoz** [comment]
well the author can do that

**Comment by sveN295** [comment]
@jackpoz Just tested it all again, everything still works as advertised. So i'd say it's ready to be merged. Only issue I had was that in one try, the Incarcerators didn't respawn. In the other tries they did. Maybe it has to do with https://github.com/TrinityCore/TrinityCore/issues/24421?

**Comment by illfated** [comment]
Sounds plausible, manual respawning does not always work.

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**